From c50cc24690938db53cd91ae9ff2fa0958693f80d Mon Sep 17 00:00:00 2001
From: Alex Deucher <alexander.deucher@amd.com>
Date: Tue, 14 Feb 2012 16:32:17 +0000
Subject: radeon: add TN surface support

Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
---
diff --git a/radeon/r600_pci_ids.h b/radeon/r600_pci_ids.h
index 0ffb741..8fbd749 100644
--- a/radeon/r600_pci_ids.h
+++ b/radeon/r600_pci_ids.h
@@ -269,3 +269,11 @@ CHIPSET(0x6768, CAICOS_6768, CAICOS)
 CHIPSET(0x6770, CAICOS_6770, CAICOS)
 CHIPSET(0x6778, CAICOS_6778, CAICOS)
 CHIPSET(0x6779, CAICOS_6779, CAICOS)
+
+CHIPSET(0x9900, ARUBA_9900, ARUBA)
+CHIPSET(0x9901, ARUBA_9901, ARUBA)
+CHIPSET(0x9903, ARUBA_9903, ARUBA)
+CHIPSET(0x9904, ARUBA_9904, ARUBA)
+CHIPSET(0x990F, ARUBA_990F, ARUBA)
+CHIPSET(0x9990, ARUBA_9990, ARUBA)
+CHIPSET(0x9991, ARUBA_9991, ARUBA)
diff --git a/radeon/radeon_surface.c b/radeon/radeon_surface.c
index 94224f5..6303c47 100644
--- a/radeon/radeon_surface.c
+++ b/radeon/radeon_surface.c
@@ -68,6 +68,7 @@ enum radeon_family {
     CHIP_TURKS,
     CHIP_CAICOS,
     CHIP_CAYMAN,
+    CHIP_ARUBA,
     CHIP_LAST,
 };
 
--
cgit v0.9.0.2-2-gbebe
