diff -up libdrm-20080930/libdrm/Makefile.am.fixlink libdrm-20080930/libdrm/Makefile.am
--- libdrm-20080930/libdrm/Makefile.am.fixlink	2008-10-01 11:18:36.000000000 +1000
+++ libdrm-20080930/libdrm/Makefile.am	2008-10-01 11:18:44.000000000 +1000
@@ -27,7 +27,6 @@ libdrm_la_LDFLAGS = -version-number 2:3:
 AM_CFLAGS = -I$(top_srcdir)/shared-core
 libdrm_la_SOURCES = xf86drm.c xf86drmHash.c xf86drmRandom.c xf86drmSL.c \
 	xf86drmMode.c
-libdrm_la_LIBADD = intel/libdrm_intel.la
 
 libdrmincludedir = ${includedir}
 libdrminclude_HEADERS = xf86drm.h xf86drmMode.h
