diff --git a/libdrm/radeon/radeon_bo_gem.c b/libdrm/radeon/radeon_bo_gem.c
index 932afeb..6d2fc61 100644
--- a/libdrm/radeon/radeon_bo_gem.c
+++ b/libdrm/radeon/radeon_bo_gem.c
@@ -221,3 +221,9 @@ void radeon_bo_manager_gem_dtor(struct radeon_bo_manager *bom)
     }
     free(bomg);
 }
+
+uint32_t radeon_gem_name_bo(struct radeon_bo *bo)
+{
+    struct radeon_bo_gem *bo_gem = (struct radeon_bo_gem*)bo;
+    return bo_gem->name;
+}
diff --git a/libdrm/radeon/radeon_bo_gem.h b/libdrm/radeon/radeon_bo_gem.h
index c0f68e6..980a6a4 100644
--- a/libdrm/radeon/radeon_bo_gem.h
+++ b/libdrm/radeon/radeon_bo_gem.h
@@ -37,4 +37,5 @@
 struct radeon_bo_manager *radeon_bo_manager_gem_ctor(int fd);
 void radeon_bo_manager_gem_dtor(struct radeon_bo_manager *bom);
 
+uint32_t radeon_gem_name_bo(struct radeon_bo *bo);
 #endif
